### ISSUE TITLE
[question] Is it possible to provide a user defined property file to the MSBuild helper ?

### ISSUE THREAD

**Issue by cedriclaplace** [conan-io/conan#6053]
I would like to know if there is a way I can tell Conan MSBuild helper to use a user defined property file to override some project settings or is there a way to control what conan generates for its own file ?

We have a bunch of Visual Studio solutions/projects in our company, each of them defining their own settings. We would like to homogenize several things and we do not want to update manually all sln/vcxproj files. To achieve that, we would like to force some project settings with a property file which can be templated to handle some specific cases. A concrete example is to define compiler settings.

I manage to give my own property file by creating a new class inheriting from the MSbuild helper and where I overrides the build method to use this file.

Instead of generating a new property file, do you think it'll be possible to use the one given through property_file_name parameter if it points to an existing property file ? 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by lasote** [comment]
Thanks for your question. Would be possible to know how that file looks? I would like to learn which things are you adjusting to and if it makes sense to consider more options to the build helper.

> I manage to give my own property file by creating a new class inheriting from the MSbuild helper and where I overrides the build method to use this file.

Nice to hear that you can customize Conan to fit your needs!

> Instead of generating a new property file, do you think it'll be possible to use the one given through property_file_name parameter if it points to an existing property file ?

Not sure if that would be the best approach, maybe it is a bit messy. Maybe it might be possible to pass several files on a list? Where is the file you are "injecting"? Are you exporting it with the recipe or is it external to the recipe or...?

Thanks

**Comment by cedriclaplace** [comment]
Thanks for the quick reply.

For now, the file is quite simple, we only defined compiler settings (see below). The file is embedded in a PythonRequire module. The project recipe inherits from a base ConanFile class, also provided through this PythonRequire module, which has a build method using this file.

We have tried 2 approches:
1/ The first one was to provide an already existing file.
2/ The second one was to use conan profil as you do for the runtime library flag.

The first option is like a short term option which fits our needs or now. The second one is probably a better approch.

The first one is the easiest approch but I agree it is probably not the best approch. We are looking to use CMake later and using this kind of file will probably require more work to configure CMake. As said, we have a PythonRequire module providing several utilities for our recipes. This package provides the property files for several build configurations (Release, Debug, ...). Our base ConanFile class has its own build method which retrieve the file to use like this:

```python
def get_build_property_file(conanfile):
  """
  Returns the msbuild property file to use to build the project.
  """
  filepath_pattern = '{basepath}\\resource\\vs\\properties\\{toolset}_{arch}_{buildtype}.props'
  return filepath_pattern.format(basepath=conanfile.python_requires['conan_common'].exports_folder,
                                 toolset=conanfile.settings.compiler.toolset,
                                 arch=conanfile.settings.arch,
                                 buildtype=str(conanfile.settings.build_type).lower())
```

This is an example of a property file we are used for testings:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ImportGroup Label="PropertySheets" />
  <PropertyGroup Label="UserMacros" />
  <ItemDefinitionGroup>
    <!-- Compiling options -->
    <ClCompile>
      <!-- Warnings/Error management -->
      <TreatWarningAsError>true</TreatWarningAsError>
      <WarningLevel>EnableAllWarnings</WarningLevel>
      <SDLCheck>true</SDLCheck>
      <!-- Runtime checks -->
      <BasicRuntimeChecks>EnableFastChecks</BasicRuntimeChecks>
      <BufferSecurityCheck>true</BufferSecurityCheck>
      <RuntimeTypeInfo>true</RuntimeTypeInfo>
      <SmallerTypeCheck>true</SmallerTypeCheck>
      <!-- Optimizations -->
      <FavorSizeOrSpeed>Neither</FavorSizeOrSpeed>
      <InlineFunctionExpansion>Disabled</InlineFunctionExpansion>
      <IntrinsicFunctions>true</IntrinsicFunctions>
      <Optimization>Disabled</Optimization>
      <StringPooling>true</StringPooling>
      <WholeProgramOptimization>false</WholeProgramOptimization>
      <!-- Debugging information -->
      <DebugInformationFormat>EditAndContinue</DebugInformationFormat>
      <!-- Miscellaneous -->
      <ExceptionHandling>Sync</ExceptionHandling>
      <FunctionLevelLinking>true</FunctionLevelLinking>
      <MinimalRebuild>true</MinimalRebuild>
      <MultiProcessorCompilation>true</MultiProcessorCompilation>
      <RuntimeLibrary>MultiThreadedDebugDLL</RuntimeLibrary>
      <!-- Preprocessor definitions -->
      <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
    </ClCompile>
    <!-- Linking options -->
    <PropertyGroup>
      <LinkIncremental>true</LinkIncremental>
    </PropertyGroup>
    <Link>
      <EnableCOMDATFolding>true</EnableCOMDATFolding>
      <GenerateDebugInformation>true</GenerateDebugInformation>
      <LinkTimeCodeGeneration>UseFastLinkTimeCodeGeneration</LinkTimeCodeGeneration>
    </Link>
  </ItemDefinitionGroup>
  <ItemGroup />
</Project>
```

The second one offers more flexibility I think but require more work. We had to define possible options and handle it manually, deal with compiler version, etc... We tried to put them into the settings.compiler section or in the options section, that works in both case. We had to deal with option declaration in settings.yml or conan recipe. This also impact the dependency lookup so we need to remove meaningless options/settings. Maybe I miss something to simplify this. At this point I don't really know which section is the best one.
In both cases, our CustomMSBuild helper reads settings (or options) to construct the property file. We simply redefined the internal MSBuild._get_props_file_contents() method (nothing special done here, we do the same thing like you do for runtime library option). We are looking to configure CMake with this but it's not working for now.

Options are declared like this in the profile:
```text
[options]
cxx_flag_basic_runtime_checks=RTC1
cxx_flag_buffer_security_check=True
cxx_flag_debug_information_format=ZI
cxx_flag_exception_handling=EHsc
cxx_flag_favor_size_or_speed=Neither
cxx_flag_function_level_linking=True
cxx_flag_inline_function_expansion=Ob0
cxx_flag_intrinsic_functions=False
cxx_flag_minimal_rebuild=True
cxx_flag_multi_processor_compilation=True
cxx_flag_optimization_level=Od
cxx_flag_runtime_library=MDd
cxx_flag_runtime_typeinfo=True
cxx_flag_smaller_type_check=True
cxx_flag_sdl_check=True
cxx_flag_treat_warning_as_error=True
cxx_flag_warning_level=Wall
cxx_flag_whole_program_optimization=True
linker_flag_enable_comdat_folding=True
linker_flag_debug_information=True
linker_flag_link_incremental=False
linker_flag_optimize_references=True
```

Thanks!

**Comment by lasote** [comment]
A lot of thanks for your explanation. I agree with you, both approaches are totally valid, the second one would be a bit more complex to develop but more powerful. 

So, how are you injecting the prop files to call the MSBuild? Are you adding additional `/p:ForceImportBeforeCppTargets={basepath}\\resource\\vs\\properties\\{toolset}_{arch}_{buildtype}.props` ?

Let's try to figure out what is the best way to inject more prop files.

**Comment by cedriclaplace** [comment]
We only have one property file in both approaches so we have only one ForceImportBeforeCppTargets parameter defined on the command line. In the first approach, we use a default file and the user can specify another one if needed. In the second approach, the property file is generated based on settings. The second approach does not need to have multiple imports but for the first one it should be preferable to ensure default settings.

To answer your question about multiple property files, I have tried a couple of things:
1. I have tried to pass 2 property files using 2 ForceImportBeforeCppTargets property parameters. It's not working, the second one overrides completly the first one and settings defined in the first one are lost. Maybe I am doing something wrong but it looks like being the wrong way.
2. I have tried to pass a list of files through a single ForceImportBeforeCppTargets parameter. Using a list of files with a semi-colon as separator seems to work. Settings defined in the second file overrides settings in the first file and settings defined in the first file seems untouched if they are not defined in the second file.

For my test I am using 2 property files. The first one is defined in my previous comment. The second one overrides warning management settings as follow:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ImportGroup Label="PropertySheets" />
  <PropertyGroup Label="UserMacros" />
  <ItemDefinitionGroup>
    <!-- Compiling options -->
    <ClCompile>
      <!-- Warnings/Error management -->
      <TreatWarningAsError>false</TreatWarningAsError>
      <WarningLevel>TurnOffAllWarnings</WarningLevel>
    </ClCompile>
  </ItemDefinitionGroup>
  <ItemGroup />
</Project>
``` 

According to these files, we are expecting the following command line parameters for CL:
From the first file:
```text
TreatWarningAsError=WX
WarningLevel=Wall
BasicRuntimeChecks=
BufferSecurityCheck=GS
RuntimeTypeInfo=GR-
SmallerTypeCheck=
FavorSizeOrSpeed=Ot
InlineFunctionExpansion=Ob2
IntrinsicFunctions=Oi
Optimization=O2
StringPooling=GF
WholeProgramOptimization=GL
DebugInformationFormat=Zi
ExceptionHandling=EHsc
FunctionLevelLinking=Gy
MinimalRebuild=Gm-
MultiProcessorCompilation=MP
RuntimeLibrary=MD
```

From the second file
```text
TreatWarningAsError=WX-
WarningLevel=W0
```

Our CustomMSBuild helper overloads the build method to support a list of property files, if None is given, the default behavior is used (the ConanFile is responsible to provides property files to use). It also overloads the get_command method to properly format the command line:
```python
  def get_command(self, project_file, props_filepaths=None, targets=None, upgrade_project=True,
                  build_type=None, arch=None, parallel=True, toolset=None, platforms=None,
                  use_env=False, properties=None, output_binary_log=None, verbosity=None):
    ...
    if props_filepaths and len(props_filepaths) > 0:
        command.append('/p:ForceImportBeforeCppTargets="%s"' % ';'.join(props_filepaths))
    ...
```

This is the generated commands:
- Case 1: Two ForceImportBeforeCppTargets parameters
```bat
msbuild.exe /m:8 /p:Configuration=Release /p:UseEnv=true /p:Platform=x64 /p:PlatformToolset=v140 /p:ForceImportBeforeCppTargets=E:\resource\vs\properties\v140_x86_64_release.props /p:ForceImportBeforeCppTargets=E:\resource\vs\properties\v140_x86_64_nowarns.props /target:MyProject /verbosity:diagnostic E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215/Solutions\myproject.sln
CL.exe /c /Zi /nologo /W0 /WX- /MP /O2 /Oi /GL /D WIN32 /D NDEBUG /D _WINDOWS /D _USRDLL /D MYPROJECT_EXPORTS /D _WINDLL /D _AFXDLL /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215\MyProject\Release\\" /Fd"E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215\MyProject\Release\MyProject_mmti.pdb" /Gd /TP /errorReport:queue main.cpp (TaskId:20)
```

We can see here that all informations coming from the first property files are not set on the CL command line. We have our warning management settings set (/W0 /WX-) but none of the non defaulted settings are missing (we do not have InlineFunctionExpansion (Ob2) set for example).

- Case 2: List of files through one ForceImportBeforeCppTargets parameter:
```bat
msbuild.exe /m:8 /p:Configuration=Release /p:UseEnv=true /p:Platform=x64 /p:PlatformToolset=v140 /p:ForceImportBeforeCppTargets=E:\resource\vs\properties\v140_x86_64_release.props;E:\resource\vs\properties\v140_x86_64_nowarns.props /target:MyProject /verbosity:diagnostic E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215/Solutions\myproject.sln
CL.exe /c /Zi /nologo /W0 /WX- /MP /O2 /Ob2 /Oi /Ot /GL /D WIN32 /D NDEBUG /D _WINDOWS /D _USRDLL /D MYPROJECT_EXPORTS /D _WINDLL /D _AFXDLL /GF /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR- /Fo"E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215\MyProject\Release\\" /Fd"E:\.conan\data\myproject\1.0.0\tester\alpha\build\e603527b7bbbe218ffd713b974c3c8a2ef9d2215\MyProject\Release\MyProject_mmti.pdb" /Gd /TP /errorReport:queue main.cpp (TaskId:20)
```

We can see here that we have our informations coming from the first property file properly defined and the second one overriding warning management settings (We have /W0 /WX- instead of /WX /Wall).

The second solution seems to work but I'm not sure about the correctness of this solution. Let me know what you think about this or if you have another option.

Thanks.

**Comment by lasote** [comment]
Thanks so much for the info, we will take a look and thing the best approach.

**Comment by memsharded** [comment]
I think this is very, very aligned with the ongoing effort in the "toolchains" https://github.com/conan-io/conan/issues/5737

Which has already some work, into the "CMake" toolchain generator: https://github.com/conan-io/conan/pull/5919

I am going to re-read this question now, but sounds like something that should be supported by toolchains.

**Comment by memsharded** [comment]
https://stackoverflow.com/questions/3924404/msbuild-custom-after-microsoft-common-targets-for-native-c-projects-in-vs2010 they use this syntax, concatenating the ``ForceImportAfterCppTargets`` too:

```
msbuild CppTest01.vcxproj 
    /p:ForceImportBeforeCppTargets="C:\Temp\_NET\ThrowAway\CppTest01\CppTest01\Before.proj";
    ForceImportAfterCppTargets="C:\Temp\_NET\ThrowAway\CppTest01\CppTest01\After.proj"
```

I am checking if it is possible to make a ``.props`` file to include another ``.props`` file, but it is very difficult with such a poor documentation of MSBuild.

**Comment by memsharded** [comment]
I am proposing to add an extra argument to the methods, so it is easy to define such an extra file. 

Check the approach here: https://github.com/conan-io/conan/pull/6374, lets see if it moves forward. You can also try running from sources from my branch in case you want to test the feature live in your computer too, that would be useful if possible.

**Comment by cedriclaplace** [comment]
First of all, thank you for the improvement.

I took a look at your feature and I tried it. It's quite similar of what I do with my own implementation. For us it's ok, we can do what we want and it does the job.

Maybe a slight improvement: Instead of taking one property file, allow the user to provide a list of property files to use as we discussed before. We don't really need this feature but we can imagine to have our build system providing a default property file and a user providing a project specific one. As a workaround, we can easilly imagine to have something generating a property file properly and pass a single property file to conan.

We can imagine to have something like that to handle both situation:
```python
if props_file_path or user_property_file_name:
    if isinstance(user_property_file_name, list):
        props_files = (props_file_path, *user_property_file_name)
    else:
        props_files = (props_file_path, user_property_file_name)
    paths = [os.path.abspath(p) for p in props_files if p]
    paths = ";".join(paths)
    command.append('/p:ForceImportBeforeCppTargets="%s"' % paths)
```
